cli/error.go

```diff
@@ -6,16 +6,22 @@ import (
 
 // StatusError reports an unsuccessful exit by a command.
 type StatusError struct {
-	Status     string
+	Status     error
```

Comment:
I'll think a bit about a solution for this, but I don't think we should concern ourselves too much with externals importing this particular package.

Reply:
>  but I don't think we should concern ourselves too much with externals importing this particular package.

For any reason in particular?

Reply:
AFAIK we have no SLA stating that internal packages in the CLI should never break due to external usage. I'd take the side of our duties are to Docker and Docker's users before a 3rd party using our code.

Reply:
Sure, but a more nuanced take on this would consider the fact that we greatly benefit from 3rd party users integrating with and using our packages, and that we lose nothing by taking that into consideration before shipping breaking changes. And that's aside from the fact that it's just good stewardship of open source projects like these to consider the wide impacts of a breaking API change.

Reply:
Also,
> internal packages

This package is specifically not "internal". In Go, internal packages not meant to be depended on are generally placed under `internal/`, but we didn't do this. If we had, this wouldn't be a discussion. However, since we did export this, we should try to adhere to the implicit contract.

Reply:
I don't agree. If this was an API client, then sure. If this was a behavior change of a feature of the CLI, then sure. But this is an error type used within the CLI, so I don't agree with your assessment. Again our duties are first towards Docker and Docker's users. Since this change is beneficial to both it outweighs any risk we perceive to third parties.

We cannot allow ourselves to be blocked by, in my opinion, an inconsequential risk. Breaking the error type just simply means the third party has to change it when they update. 

The impact is also anecdotal and doesn't necessarily convey the actual effect since the GitHub search only tells half the story. We honestly don't even know if there will be a risk to those third party projects as we don't keep track of them :shrug:

Reply:
https://www.hyrumslaw.com/

Reply:
It's an exported type, so it's part of the API. That isn't an opinion to agree or disagree on, it's a fact.

> Since this change is beneficial to both it outweighs any risk we perceive to third parties.

I'll reiterate since maybe I wasn't clear:
> It might be better to fix this without changing cli.StatusError first, and do this change separately, unless we absolutely need it to fix this.

I'm not saying we can't ever change the API, we often do. I'm saying that breaking the API has consequences: what kind of releases we can ship it in (major vs. minor), more work for consumers, etc., and **if we can do fix this now without changing the API and hold off on the change we should**.

This means that if there is an option that allows us to get that benefit to our users without the negatives, then we should prefer that over the alternatives. Not that we would
> allow ourselves to be blocked by, in my opinion, an inconsequential risk.

Reply:
I understand where you are coming from, but I want to clarify my position. Docker as an Organization does not have an agreement with third parties to consume the CLI code (afaik). 

Having an implicit SLA through Go's module specification for the CLI seems counter-intuitive to me as the CLI is not built as a library to be consumed.

With this implicit SLA contributing anything, however small (as in this case), will always get blocked or pushed out to a major release, which I see as counterproductive.

Reply:
> With this implicit SLA contributing anything, however small (as in this case), will always get blocked or pushed out to a major release, which I see as counterproductive.

This is simply not true. There are various ways to contribute changes that don't break semver API contracts.

> Having an implicit SLA through Go's module specification for the CLI seems counter-intuitive to me as the CLI is not built as a library to be consumed.

So your proposal is that we simply disregard Semver/Go Modules, and ship breaking API changes in minor/patch releases? If everyone started doing this, everything in the Go Ecosystem would break due to the toolchain no longer being able to depend on guarantees such as these when selecting versions, updating any dependencies would become a chore, pipelines would break overnight, etc.

That's discounting the fact that, again, we benefit (and want) people to build things using our packages, and doing that requires building trust on the stability of those packages. Some packages we can (and should) move/deprecate (on major releases) to `/internal` if we don't want them being used (and we have done this), but as long as they're explicitly exported we must uphold the expectations we subscribed to when we shipped exported packages. Breaking this wouldn't even be okay from an internal API consumer's point of view.